controllers/topology/reconcile_state_test.go

```diff
@@ -0,0 +1,361 @@
+			for k, v := range expectedSpec {
+				g.Expect(cloneSpec).To(HaveKeyWithValue(k, v))
+			}
+		})
+	}
+}
+
+type referencedObjectsCompatibilityTestCase struct {
+	name    string
+	current *unstructured.Unstructured
+	desired *unstructured.Unstructured
+	wantErr bool
+}
+
+var referencedObjectsCompatibilityTestCases = []referencedObjectsCompatibilityTestCase{
+	{
+		name: "Fails if group changes",
+		current: &unstructured.Unstructured{
+			Object: map[string]interface{}{
+				"apiVersion": "foo/v1alpha4",
+			},
+		},
+		desired: &unstructured.Unstructured{
+			Object: map[string]interface{}{
+				"apiVersion": "bar/v1alpha4",
+			},
+		},
+		wantErr: true,
+	},
+	{
+		name: "Fails if kind changes",
+		current: &unstructured.Unstructured{
+			Object: map[string]interface{}{
+				"kind": "foo",
+			},
+		},
+		desired: &unstructured.Unstructured{
+			Object: map[string]interface{}{
+				"kind": "bar",
+			},
+		},
+		wantErr: true,
+	},
+	{
+		name: "Pass if gvk remains the same",
+		current: &unstructured.Unstructured{
+			Object: map[string]interface{}{
+				"apiVersion": "infrastructure.cluster.x-k8s.io/foo",
+				"kind":       "foo",
+			},
+		},
+		desired: &unstructured.Unstructured{
+			Object: map[string]interface{}{
+				"apiVersion": "infrastructure.cluster.x-k8s.io/foo",
+				"kind":       "foo",
+			},
+		},
+		wantErr: false,
+	},
+	{
+		name: "Pass if version changes but group and kind remains the same",
+		current: &unstructured.Unstructured{
+			Object: map[string]interface{}{
+				"apiVersion": "infrastructure.cluster.x-k8s.io/foo",
+				"kind":       "foo",
+			},
+		},
+		desired: &unstructured.Unstructured{
+			Object: map[string]interface{}{
+				"apiVersion": "infrastructure.cluster.x-k8s.io/bar",
+				"kind":       "foo",
+			},
+		},
+		wantErr: false,
+	},
+	{
+		name: "Fails if namespace changes",
+		current: &unstructured.Unstructured{
+			Object: map[string]interface{}{
+				"metadata": map[string]interface{}{
+					"namespace": "foo",
+				},
+			},
+		},
+		desired: &unstructured.Unstructured{
+			Object: map[string]interface{}{
+				"metadata": map[string]interface{}{
+					"namespace": "bar",
+				},
+			},
+		},
+		wantErr: true,
+	},
+}
+
+func TestCheckReferencedObjectsAreCompatible(t *testing.T) {
+	for _, tt := range referencedObjectsCompatibilityTestCases {
+		t.Run(tt.name, func(t *testing.T) {
+			g := NewWithT(t)
+
+			err := checkReferencedObjectsAreCompatible(tt.current, tt.desired)
+			if tt.wantErr {
+				g.Expect(err).To(HaveOccurred())
+				return
+			}
+			g.Expect(err).ToNot(HaveOccurred())
+		})
+	}
+}
+
+func TestCheckReferencedObjectsAreStrictlyCompatible(t *testing.T) {
+	referencedObjectsStrictCompatibilityTestCases := append(referencedObjectsCompatibilityTestCases, []referencedObjectsCompatibilityTestCase{
+		{
+			name: "Fails if name changes",
+			current: &unstructured.Unstructured{
+				Object: map[string]interface{}{
+					"metadata": map[string]interface{}{
+						"namespace": "foo",
```

Comment:
```suggestion
						"name": "foo",
```
I think, same in the other cases.